packages/headless/src/app/engine-configuration-options.ts

```diff
@@ -0,0 +1,109 @@
+import {PreprocessRequest} from '../api/preprocess-request';
+
+/**
+ * The global headless engine configuration options.
+ */
+export interface EngineConfigurationOptions {
+  /**
+   * The unique identifier of the target Coveo Cloud organization (e.g., `mycoveocloudorganizationg8tp8wu3`)
+   */
+  organizationId: string;
+  /**
+   * The access token to use to authenticate requests against the Coveo Cloud endpoints. Typically, this will be an API key or search token that grants the privileges to execute queries and push usage analytics data in the target Coveo Cloud organization.
+   */
+  accessToken: string;
+  /**
+   * A function that fetches a new access token. The function must return a Promise that resolves to a string (the new access token).
+   */
+  renewAccessToken?: () => Promise<string>;
+  /**
+   * Allows for augmenting a Platform request before it is sent.
+   * @param request Request to be augmented
+   * @param clientOrigin The origin of the client, can be "analyticsFetch", "analyticsBeacon" or "searchApiFetch"
+   *
+   * @returns Augmented request
+   */
+  preprocessRequest?: PreprocessRequest;
+  /**
+   * The Plaform URL to use. (e.g., https://platform.cloud.coveo.com)
+   * The platformUrl() helper method can be useful to know what url is available.
+   */
+  platformUrl?: string;
+  /**
+   * Allows configuring options related to analytics.
+   */
+  analytics?: {
+    /**
+     * Specifies if usage analytics tracking should be enabled.
+     *
+     * By default, all analytics events will be logged.
+     */
+    enabled?: boolean;
+    /**
+     * Origin level 2 is a usage analytics event metadata whose value should typically be the name/identifier of the tab from which the usage analytics event originates.
+     *
+     * When logging a Search usage analytics event, originLevel2 should always be set to the same value as the corresponding tab (parameter) Search API query parameter so Coveo Machine Learning models function properly, and usage analytics reports and dashboards are coherent.
+     *
+     * This value is optional, and will automatically try to resolve itself from the tab search parameter.
+     */
+    originLevel2?: string;
+    /**
+     * Origin level 3 is a usage analytics event metadata whose value should typically be the URL of the page that linked to the search interface that’s making the request.
+     *
+     * When logging a Search usage analytics event, originLevel3 should always be set to the same value as the corresponding referrer Search API query parameter so usage analytics reports and dashboards are coherent.
+     *
+     * This value is optional, and will automatically try to resolve itself from the referrer search parameter.
+     */
+    originLevel3?: string;
+    /**
+     * analyticsClientMiddleware allows to hook into an analytics event payload before it is sent to the Coveo platform.
+     */
+    analyticsClientMiddleware?: AnalyticsClientSendEventHook;
```

Comment:
I see the [analyticsMiddleware](https://github.com/coveo/ui-kit/blob/master/packages/headless/src/app/analytics-middleware.ts) currently handles search, recommendation, and commerce stuff. I guess it would be a good idea to split it and combine the middlewares as needed depending on the use case.

Reply:
Indeed, though I think the savings are smaller here since it's just a chain of `if` statements. Still, it's a valid point and there might be something we can do better.